providers/implementations/keymgmt/dh_kmgmt.c

```diff
@@ -357,8 +344,250 @@ static int dh_validate(void *keydata, int selection)
+    };
+    return settable;
+}
+
+static int dh_gencb(int p, int n, BN_GENCB *cb)
+{
+    struct dh_gen_ctx *gctx = BN_GENCB_get_arg(cb);
+    OSSL_PARAM params[] = { OSSL_PARAM_END, OSSL_PARAM_END, OSSL_PARAM_END };
+
+    params[0] = OSSL_PARAM_construct_int(OSSL_GEN_PARAM_POTENTIAL, &p);
+    params[1] = OSSL_PARAM_construct_int(OSSL_GEN_PARAM_ITERATION, &n);
+
+    return gctx->cb(params, gctx->cbarg);
+}
+
+static void *dh_gen(void *genctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    struct dh_gen_ctx *gctx = genctx;
+    DH *dh = NULL;
+    BN_GENCB *gencb = NULL;
+    int ret = 0, named_group = 0;
+    FFC_PARAMS *ffc;
+
+    if (gctx == NULL)
+        return NULL;
+
+    /* For parameter generation - If there is a group name just create it */
+    if (gctx->group_nid != NID_undef
+        || gctx->type == DH_PARAMGEN_TYPE_GROUP) {
+        /* Select a named group if there is not one already */
+        if (gctx->group_nid == NID_undef) {
+            gctx->group_nid = dh_get_named_group_uid_from_size(gctx->pbits);
+            if (gctx->group_nid == NID_undef)
+                return NULL;
+        }
+        dh = dh_new_by_nid_with_libctx(gctx->libctx, gctx->group_nid);
+        if (dh == NULL)
+            return NULL;
+        named_group = 1;
+    } else {
+        dh = dh_new_with_libctx(gctx->libctx);
+        if (dh == NULL)
+            return NULL;
+    }
+
+    gctx->cb = osslcb;
+    gctx->cbarg = cbarg;
+    gencb = BN_GENCB_new();
+    if (gencb != NULL)
+        BN_GENCB_set(gencb, dh_gencb, genctx);
+
+    ffc = dh_get0_params(dh);
+    if (!named_group) {
+        /* Copy the template value if one was passed */
+        if (gctx->ffc_params != NULL
+            && !ffc_params_copy(ffc, gctx->ffc_params))
+            goto end;
+
+        /* Otherwise parameter generation is based on the generation type */
+        if (!ffc_params_set_seed(ffc, gctx->seed, gctx->seedlen))
+            goto end;
+        if (gctx->gindex != -1)
+            ffc_params_set_gindex(ffc, gctx->gindex);
+
+        if ((gctx->selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0) {
+            /*
+             * NOTE: The old safe prime generator code is not used in fips mode,
+             * (i.e internally it ignores the generator and chooses a named
+             * group based on pbits.
+             */
+            if (gctx->type == DH_PARAMGEN_TYPE_GENERATOR)
+                ret = DH_generate_parameters_ex(dh, gctx->pbits,
+                                                gctx->generator, gencb);
+            else
+                ret = dh_generate_ffc_parameters(dh, gctx->type, gctx->pbits,
+                                                 gctx->qbits, gctx->md, gencb);
+            if (ret <= 0)
+                goto end;
+        }
+    }
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0)
+        ret = DH_generate_key(dh);
+end:
+    if (!ret) {
+        fprintf(stdout, "FAILED TO GENERATE DH KEY");
```

Comment:
Debug code?

Reply:
removed - I was running some tests and seeing if it hit this anywhere,,